### PR TITLE
Improve nvda helper by compiling with C++20 and disabling the /permissive flag

### PR DESCRIPTION
**Pull request by LeonarddeR**
### Link to issue number:
None

### Summary of the issue:
NVDAHelper was compiled with C++17, it could be compiled with C++20

1. Visual Studio has support for C++20.
2. WinRT is C++20 (#11768).
3. Compiling with C++20 ensures we're more conformant.
   - Passing string literals to parameters expecting `wchar_t*` or `BSTR` is no longer supported. This ensures safe code. See [here](https://docs.microsoft.com/en-us/cpp/build/reference/zc-strictstrings-disable-string-literal-type-conversion?view=msvc-170) for a rationale.
   - The`/permissive-` flag (note the minus symbol disabling permissive mode), ensuring conformance in more areas. See [here](https://docs.microsoft.com/en-us/cpp/build/reference/permissive-standards-conformance?view=msvc-170)

### Description of how this pull request fixes the issue:
- Change build params so `nvdaHelper` is built with C++20 and disable permissive mode.
- Fix errors (`wchar_t*` or `BSTR` from string literal)

### Testing strategy:
The following areas of code have been touched and need testing:

* UIA custom automation types (comments in Excel with UIA and Windows 11)
* Adobe reader virtual buffers
* Gecko virtual buffers
* MSHTML virtual buffers
* Webkit virtual buffers

Other changes are mainly changes from `char*` to `const char*`, so unlikely to cause any harm.

I did some quick tests and didn't find anything crucial, probably best to test this in Alpha.

### Known issues with pull request:
None

### Change log entries:
For Developers
- NVDAHelper is now C++20 and built with the /permissive- compiler flag, which ensures better conformance to C++ standards.
- 
### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English

### CODE REVIEW

**Comment by leonardder** [comment]
I'm pretty sure the system test failure is not related to this pr.

**Comment by feerrenrut** [comment]
The last build failed. I'm going to convert this to a draft for now.

**Comment by leonardder** [comment]
The build issue should have been fixed. It was due to a bad merge.

**Comment by feerrenrut** [comment]
Unless there is some urgency on this that I'm not aware of, I think it would be better to merge this in the new dev cycle. Applying 2022.2 milestone.

**Comment by seanbudd** [comment]
We plan to merge this soon for 2022.2, and are ready to revert if there are issues

**Comment by leonardder** [comment]
It looks like the UIA remote library doesn't like C++20 that much. I need to dive into this some more.

File: nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp [hunk 1]
```diff
@@ -89,7 +89,7 @@ CComPtr<IAccessible2> GeckoVBufBackend_t::getRelationElement(
 const wchar_t EMBEDDED_OBJ_CHAR = 0xFFFC;
 // Always render a space for "empty" / metadata only
 // text leaf nodes so the user can access them.
-constexpr wchar_t* EMPTY_TEXT_NODE {L" "};
+const wchar_t* EMPTY_TEXT_NODE {L" "};
```
Comment (leonardder) [inline nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp:92]:
```suggestion
constexpr const wchar_t* EMPTY_TEXT_NODE[]{L" "};
```

**Comment by feerrenrut** [comment]
>It looks like the UIA remote library doesn't like C++20 that much. I need to dive into this some more.

Did you resolve these issues?

**Comment by leonardder** [comment]
This is now ready for review. Probably want to merge this early.

File: nvdaHelper/archBuild_sconscript [hunk 2]
```diff
@@ -144,13 +137,14 @@ if 'RTC' in debug:
 env.Append(PDB='${TARGET}.pdb')
 env.Append(LINKFLAGS='/OPT:REF') #having symbols usually turns this off but we have no need for unused symbols
 
-# Ignoring Flake8 F821: 'undefined name' due to nonstandard SCons import
-win10env = env.Clone()  # noqa: F821
-# Swap C++17 for C++20 for co-routine support, used by C++/WinRT.
-win10env.Replace(CCFLAGS=[
+env.Append(CCFLAGS=[
 	'/std:c++20',
```
[the remark this answers is not in the document]
Comment (leonardder) [reply nvdaHelper/archBuild_sconscript:141]: I'm confused, what's wrong in your opinion?
As C++20 is the default as of this pr, we no longer need a specific environment for Win10.

[the remark this answers is not in the document]
Comment (feerrenrut) [reply nvdaHelper/archBuild_sconscript:141]: Sorry, looking at it again I see that it is fine. I was worried that these used to be `CCFLAGS` specific to an individual project, now changing the build environment for all.

**Review by feerrenrut — APPROVED** [review]

**Pull request merged**